File: 128.最长连续序列.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // unordered_map<int, int> m;
        // int res = 0;
        // for(auto num : nums){
        //     if(m[num] == 0){
        //         int left = m[num - 1];
        //         int right = m[num + 1];
        //         //该点的区间长度值为其左右相邻数的区间长度值＋1
        //         m[num] = left + right + 1;
        //         res = max(res , m[num]);
        //         // 更新两端点的区间长度值
        //         m[num - left] = m[num];
        //         m[num + right] = m[num];
        //     }
        // }
        // return res;

        // 把数插入set消除重复元素
        // 遍历每一个数num，查询num, num+1, .... , y 的个数curCount
        // 如果num-1存在于set中，则跳过，因为包含num的最长长度与包含num-1的最长长度一样
        // 比较res与curCount的大小
        set<int> s;
        int res = 0;
        for(auto num : nums)
            s.insert(num);
        for(auto num : nums){
            if(s.find(num - 1) == s.end()){
                int curCount = 1;
                int curNum = num;
                while(s.find(curNum + 1) != s.end()){
                    ++curCount;
                    ++curNum;
                }
                res = max(res , curCount);
            }
        }
        return res;
    }
};

File: LeetCode 热题 HOT 100/128.最长连续序列.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
    let map = new Map();
    let res = 0;
    for (let num of nums) {
        if (!map.has(num)) {
            let left = map.has(num - 1) ? map.get(num - 1) : 0;
            let right = map.has(num + 1) ? map.get(num + 1) : 0;
            let cur = left + right + 1;
            map.set(num, cur);
            map.set(num - left, cur);
            map.set(num + right, cur);
            res = Math.max(res, cur);
        }
    }
    return res;
};


var longestConsecutive = function (nums) {
    let set = new Set();
    let res = 0;
    for (let num of nums) {
        set.add(num);
    }
    for (let num of set) {
        if (!set.has(num - 1)) {
            let curCount = 1;
            let cur = num;
            while (set.has(cur + 1)) {
                ++curCount;
                ++cur;
            }
            res = Math.max(res, curCount);
        }
    }
    return res;
}